File: src/Game/Board/Piece.cpp
#include "Piece.h"

std::string Piece::toString() {
    std::string str = "";
    color == PieceColor::WHITE ? str += "W_" : str += "B_";
    switch (type) {
    case PieceType::PAWN:
        str += "Pawn";
        break;
    case PieceType::KNIGHT:
        str += "Knight";
        break;
    case PieceType::BISHOP:
        str += "Bishop";
        break;
    case PieceType::ROOK:
        str += "Rook";
        break;
    case PieceType::QUEEN:
        str += "Queen";
        break;
    case PieceType::KING:
        str += "King";
        break;
    default:
        break;
    }
    return str;
}

std::vector<Point> Piece::getPawnNormalAttacks(PieceColor _color) {
    std::vector<Point> points;
    if (_color == PieceColor::WHITE) {
        points.push_back(Point{ 1,1 });
        points.push_back(Point{ -1,1 });
    }
    else {
        points.push_back(Point{ 1,-1 });
        points.push_back(Point{ -1,-1 });
    }
    return points;
}

std::vector<Point> Piece::getNormalMoves() {
    std::vector<Point> movePoints;
    switch (type) {
    case PieceType::PAWN:
        if (color == PieceColor::WHITE)
            movePoints.push_back(Point{ 0,1 });
        else
            movePoints.push_back(Point{ 0,-1 });
        break;
    case PieceType::KNIGHT:
        movePoints.push_back(Point{ 1,  -2 });
        movePoints.push_back(Point{ -1, -2 });
        movePoints.push_back(Point{ 1,  2 });
        movePoints.push_back(Point{ -1, 2 });
        movePoints.push_back(Point{ -2, 1 });
        movePoints.push_back(Point{ -2, -1 });
        movePoints.push_back(Point{ 2,  1 });
        movePoints.push_back(Point{ 2,  -1 });
        break;
    case PieceType::BISHOP:
        for (int i = 0; i < 7; i++) {
            movePoints.push_back(Point{ i,i });
            movePoints.push_back(Point{ -i,-i });
            movePoints.push_back(Point{ -i,i });
            movePoints.push_back(Point{ i,-i });
        }
        break;
    case PieceType::ROOK:
        for (int i = 0; i < 7; i++) {
            movePoints.push_back(Point{ 0,i });
            movePoints.push_back(Point{ 0,-i });
            movePoints.push_back(Point{ i,0 });
            movePoints.push_back(Point{ -i,0 });
        }
        break;
    case PieceType::QUEEN:
        for (int i = 0; i < 7; i++) {
            movePoints.push_back(Point{ i,i });
            movePoints.push_back(Point{ -i,-i });
            movePoints.push_back(Point{ -i,i });
            movePoints.push_back(Point{ i,-i });

            movePoints.push_back(Point{ 0,i });
            movePoints.push_back(Point{ 0,-i });
            movePoints.push_back(Point{ i,0 });
            movePoints.push_back(Point{ -i,0 });
        }
        break;
    case PieceType::KING:
        movePoints.push_back(Point{ -1,-1 });
        movePoints.push_back(Point{ -1,0 });
        movePoints.push_back(Point{ 0,-1 });
        movePoints.push_back(Point{ 1,1 });
        movePoints.push_back(Point{ 1,0 });
        movePoints.push_back(Point{ 0,1 });
        movePoints.push_back(Point{ -1,1 });
        movePoints.push_back(Point{ 1,-1 });
        break;
    default:
        break;
    }

    return movePoints;
}

File: res/scripts/ui_game/piece.lua
Piece = {theme = "alpha"}

function Piece:new(o)
    o = o or {}
    o.width = o.width or 100
    o.x = o.x or 0
    o.y = o.y or 0
    o.type = o.type or PieceType.KING
    o.color = o.color or PieceColor.WHITE
    setmetatable(o, self)
    self.__index = self
    return o
end

function Piece:render()
    local img = Image(self:getImgPath())
    img:setSize(self.width, self.width)
    img:setPosition(self.x, self.y)
    img:render()
end

function Piece:getPieceFileName()
    return self:getPieceFileColorName() .. self:getPieceFilePieceName() .. ".png"
end

function Piece:getImgPath()
    return "pieces//" .. self.theme .. "//" .. self:getPieceFileName()
end

function Piece:getPieceFilePieceName()
    if self.type == PieceType.KING then
        return "K"
    elseif self.type == PieceType.ROOK then
        return "R"
    elseif self.type == PieceType.QUEEN then
        return "Q"
    elseif self.type == PieceType.BISHOP then
        return "B"
    elseif self.type == PieceType.KNIGHT then
        return "N"
    else
        return "P"
    end
end

function Piece:getPieceFileColorName()
    if self.color == PieceColor.WHITE then
        return "w"
    else
        return "b"
    end
end